src/main/java/org/jenkinsci/plugins/workflow/actions/ArgumentsAction.java

```diff
@@ -241,6 +249,7 @@ static boolean checkArgumentsLackPlaceholders(@Nonnull Map<String,Object> namedA
             if (ob instanceof NotStoredReason) {
                 return false;
             }
+            // TODO this would need to also check nested Map/List arguments
```

Comment:
Originally tried to use `checkArgumentsLackPlaceholders` to avoid even trying to instantiate `Step`s when some of the arguments were placeholders, but saw that this was not going to work anyway. For example

```groovy
archiveArtifacts '1.txt,2.txt,…,9999.txt'
```

will currently claim to be fully-formed even though the argument here is not stored.